### PR TITLE
IO runtime module for react

### PR DESCRIPTION
**Pull request by niekraaijmakers**
## Description

This PR adds an option to generate a module for react, that will make it very easy to set up server-side rendering in Adobe IO Runtime or local openwhisk for the SPA editor.

## Related Issue

https://github.com/adobe/aem-project-archetype/issues/695

## Motivation and Context

Our partners and customers are having a hard time learning / understanding and setting up the initial work for server side rendering in AdobeIO runtime or AWS lambda. Server side rendering is by nature a very complex topic and serverless even more. This will make getting started much easier, as it provides a point of "hello world" which works in IO runtime, openwisk and express with debug capabilities.

## How Has This Been Tested?

Aside from the automatic tests, I deployed the projects from the generated target folder to AEM.
I then ran express, openwhisk and adobeio runtime and checked that the server-side content came up and checked if the browser was fine.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/main/archetype/ui.frontend.react/src/server/aem-processor.functions.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+    return `${html}
+     <script type="application/json" id="__INITIAL_STATE__">
+         ${stateStr}
+     </script>`;
+}
+async function processSPA(args) {
+    const APP_ROOT_PATH = '/content/${appId}/${country}/${language}';
+    const wcmMode = args.wcmmode;
+    const isInEditor = wcmMode && wcmMode === 'EDIT' || wcmMode === 'PREVIEW';
```
Comment (sharanyavinod) [inline src/main/archetype/ui.frontend.react/src/server/aem-processor.functions.js:35]: Maybe we can reuse AuthoringUtils.isInEditor() provided here instead - https://github.com/adobe/aem-spa-page-model-manager/blob/master/src/AuthoringUtils.ts#L232

File: src/main/archetype/ui.frontend.react/actions/ssr/index.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    const refinedArgs = {
+        data: args,
+        pageRoot: args.__ow_headers['page-model-root-url'],
+        pagePath: args.__ow_path,
+        wcmmode: args.__ow_headers['wcm-mode']
+    }
+
+    try {
+        const response = await processor.processSPA(refinedArgs);
+        console.log("Response", response);
```
Comment (sharanyavinod) [inline src/main/archetype/ui.frontend.react/actions/ssr/index.js:16]: Expected log or to be removed?

File: src/main/archetype/ui.frontend.react/src/server/aem-processor.functions.js (hunk 1 above)
Comment (niekraaijmakers) [reply src/main/archetype/ui.frontend.react/src/server/aem-processor.functions.js:35]: No we cannot in this context I'm afraid :) as we are in nodejs / server

File: src/main/archetype/ui.frontend.react/actions/ssr/index.js (hunk 2 above)
Comment (niekraaijmakers) [reply src/main/archetype/ui.frontend.react/actions/ssr/index.js:16]: Good point, il tear it out

**Review by sharanyavinod — APPROVED** [review]
Looks good to me! :)

File: src/main/archetype/ui.frontend.react/actions/ssr/index.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        wcmmode: args.__ow_headers['wcm-mode']
+    }
+
+    try {
+        const response = await processor.processSPA(refinedArgs);
+        return {
+            headers: {
+                'Content-Type': 'text/html'
+            },
+            statusCode: 200,
```
[the remark this answers is not in the document]
Comment (niekraaijmakers) [reply src/main/archetype/ui.frontend.react/actions/ssr/index.js:20]: The response is merely HTML. If no error is thrown, it means it's 200. The catch block gives back 500.

File: src/main/archetype/ui.frontend.react/.env.development [hunk 4]
```diff
@@ -20,7 +20,13 @@
 # prepended to paths in the `asset-manifest.json` file)
 PUBLIC_URL=/etc.clientlibs/${appId}/clientlibs/clientlib-react/resources
 
-REACT_APP_PAGE_MODEL_PATH=/content/${appId}/en.model.json
+REACT_APP_PAGE_MODEL_PATH=/content/${appId}/${country}/${language}.model.json
```
[the remark this answers is not in the document]
Comment (niekraaijmakers) [reply src/main/archetype/ui.frontend.react/.env.development:23]: This is a separate bug indeed. We could put it in a separate PR, sure. I don't see why we shouldn't just fix it though

File: src/main/archetype/ui.apps/src/main/content/jcr_root/apps/__appId__/components/page/body.html [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<!-- this starts the SSR rendering chain !-->
+
+<div id="spa-root" data-sly-resource="${symbol_dollar}{resource @ resourceType='cq/remote/content/renderer/request/handler'}"></div>
```
[the remark this answers is not in the document]
Comment (niekraaijmakers) [reply src/main/archetype/ui.apps/src/main/content/jcr_root/apps/__appId__/components/page/body.html:3]: This file is deleted by the post groovy console script, if ssr is not enabled.

File: src/main/archetype/pom.xml [hunk 6]
```diff
@@ -64,6 +64,11 @@
         <sling.password>admin</sling.password>
         <vault.user>admin</vault.user>
         <vault.password>admin</vault.password>
+#if ($enableAdobeIoRuntime == "y")
+        <frontend-maven-plugin.version>1.6</frontend-maven-plugin.version>
```
[the remark this answers is not in the document]
Comment (niekraaijmakers) [reply src/main/archetype/pom.xml]: Right - unified it to 1 version.

File: README.md [hunk 7]
```diff
@@ -89,6 +89,7 @@ Name                      | Default        | Description
 `datalayer`               | `y`            | Activate integration with [Adobe Client Data Layer](https://docs.adobe.com/content/help/en/experience-manager-core-components/using/developing/data-layer/overview.html).
 `amp`                     | `n`            | Enable [AMP](https://docs.adobe.com/content/help/en/experience-manager-core-components/using/developing/amp.html) support for genenerated project templates.
 `enableDynamicMedia`      | `n`            | Enabled foundation DynamicMedia components in project policy settings and activates Dynamic Media features in Core Image component's policy.
+`enableAdobeIoRuntime`    | `n`            | Option to enable SSR for the front-end project (currently only supported for react)
```
[the remark this answers is not in the document]
Comment (niekraaijmakers) [reply README.md]: Valid point, made it easier: enableSSR . Because we also use express , so it's both serverless and serverfull.

**Review by habansal — APPROVED** [review]

**Review by pfauchere — APPROVED** [review]

**Pull request merged**